src/realm/object-store/shared_realm.hpp

```diff
@@ -95,7 +95,13 @@ struct RealmConfig {
     std::string path;
     BinaryData realm_data;
     // User-supplied encryption key. Must be either empty or 64 bytes.
-    std::vector<char> encryption_key;
+    OwnedBinaryData encryption_key;
+
+    // Treats empty and null encryption keys the same.
+    const char* encryption_key_ptr() const
+    {
+        return encryption_key.empty() ? nullptr : encryption_key.data();
+    }
```

Comment:
The old code was [incorrectly](https://en.cppreference.com/w/cpp/container/vector/data) assuming that `vector::data()` would always return `nullptr` when empty. While `OwnedBinaryData` _does_ have a distinct null state, given that keys must be exactly 64 bytes, it seemed best to treat the null and empty states as equivalent here.

Reply:
In part because that was the documented behavior before, it just didn't actually implement what was documented. Given that `std::vector` doesn't have a distinct null state, I'd argue that the existing code was broken and I'm fixing it at the same time as switching to a different type which happens to not need the fix.

But also because I'd really like SDK's (especially when using generated bindings) to not have to care about that distinction. As much as possible, I'd like `BinaryData` and `StringData` to always map to `ByteBuffer` and `String` types in the native language and not their nullable equivalents. I don't want every use of one of those types to force optionality into the API, especially in languages with strict null/optional checking. So far, this would be the only place where we would need that, and I don't think it is worth adding that concept just for this.

Ideally IMO, StringData and BinaryData would always be similar to `string_view` and `span<byte>` and not even have a distinct null state, and simply be possibly empty sequences of bytes. I think we probably rely too much on that internally for it to be a short term goal, but it would be nice if is at least possible on the API surface of core.

If you really want a distinct null state here, I could look into using `optional<OwnedBinaryData>`, where the contract is that if the optional is engaged, that the binary data must be 64 bytes long. But I don't see any benefit to that vs just allowing any empty `encryption_key` to mean "no encryption", regardless of what its data pointer is.

Reply:
`String.IsNullOrEmpty()` is IMO one of C#'s greatest early mistakes, and treating null as equivalent an empty state is an utter nightmare. Non-nullable types are great, but they need to actually be non-nullable. I don't hate the idea of using empty to indicate no key, but then it should be a type which doesn't also have null.